File: master.py
#!/usr/bin/env python3

import logging
import socket
import struct
import sys
import time

def broadcast_pos(sock: socket.socket, pos: float, clients: [(str, int)]) -> [int]:
    data = struct.pack('!d', pos)
    sent = []
    for client in clients:
        sent.append(sock.sendto(data, client))
    for client, written in zip(clients, sent):
        if written != 8:
            logging.error("Cannot send to '%s:%d'" % client)
    return sent

def main(argv: [str]) -> int:
    logging.basicConfig(level=logging.INFO)
    if len(argv) < 3:
        sys.stdout.write('Usage: ./master.py length client1 port1 [client2 port2 ...]\n\n')
        return 1
    logging.info("Setting length to '%s'" % argv[1])
    length = float(argv[1])
    clients = []
    for addr, port in zip(argv[2::2], argv[3::2]):
        logging.info("Adding client '%s:%s'" % (addr, port))
        clients.append((addr, int(port)))
    sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    sock.setsockopt(socket.SOL_SOCKET, socket.SO_BROADCAST, 1)
    
    while True:  # Loop playback
        start_time = time.monotonic()
        pos = 0
        broadcast_pos(sock, 0, clients)
        logging.info('Position: 0.000000000')
        while pos < length:
            time.sleep(1)
            pos = time.monotonic() - start_time
            broadcast_pos(sock, pos, clients)
            logging.info('Position: %.9f' % pos)

    return 0

if __name__ == '__main__':
    sys.exit(main(sys.argv))

File: master.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>
#include <time.h>
#include "NaiveSocketLibrary/NaiveSocketLibrary.h"

#ifdef _WIN32
    #include <windows.h>
#else
    #include <unistd.h>
#endif

void wait_sec( int seconds )
{
    #ifdef _WIN32
        Sleep( 1000 * seconds );
    #else
        sleep( seconds );
    #endif
}

void broadcast_pos(SOCKET sock, double d, SOCKADDR *peer) {
    sendto(sock, &d, sizeof(d), 0, peer, NSLEndpointV4SocketLen);
}

int main(int argc, char *argv[])
{
    static int so_broadcast = true;
	NSLInit();
    SOCKET sConnection = NSLSocket(AF_INET, SOCK_DGRAM, 0);
    int port;
    double length;
    sscanf(argv[1], "%d", &port);
    sscanf(argv[2], "%lf", &length);
    SOCKADDR *peer = NSLEndpointV4("255.255.255.255", port);
    setsockopt(sConnection, SOL_SOCKET, SO_BROADCAST, &so_broadcast, sizeof so_broadcast);

    double pos = 0;
    broadcast_pos(sConnection, pos, peer);
    puts("here");
    while (1) {
        pos += 1;
        if (pos > length) pos = 0;
        broadcast_pos(sConnection, pos, peer);
        printf("Position: %.9f\n", pos);
        wait_sec(1);
    }

    NSLCloseSocket(sConnection);
    free(peer);
    NSLEnd();
    return 0;
}
